docs/frontend-system/utility-apis/02-creating.md

```diff
@@ -0,0 +1,88 @@
+---
+id: creating
+title: Creating Utility APIs
+sidebar_label: Creating APIs
+# prettier-ignore
+description: Creating new utility APIs in your plugins and app
+---
+
+## When to make a utility API
+
+> TODO
+
+## Creating the Utility API contract
+
+The first step toward exposing a utility API is to define its TypeScript contract, and an API ref for consumers to find it via. This should be done in [your plugin's `-react` package](../../architecture-decisions/adr011-plugin-package-structure.md), so that it can be imported by other plugins. In this example, we have an Example plugin that wants to expose a utility API for performing some type of work.
+
+```tsx title="in @internal/plugin-example-react"
+import { createApiRef } from '@backstage/frontend-plugin-api';
+
+/**
+ * Performs some work.
+ * @oublic
+ */
+export interface WorkApi {
+  doWork(): Promise<void>;
+}
+
+/**
+ * The work interface for the Example plugin.
+ * @public
+ */
+export const workApiRef = createApiRef<WorkApi>({
+  id: 'plugin.example.work',
+});
+```
+
+Both of these are properly exported publicly from the package, so that consumers can reach them.
+
+The plugin itself now wants to provide this API and its default implementation, in the form of an API extension. Doing so means that when users install the Example plugin, an instance of the Work utility API will also be automatically available in their apps - both to the Example plugin itself, and to others. We do this in the main plugin package, not the `-react` package.
+
+```tsx title="in @internal/plugin-example"
+import {
+  createApiExtension,
+  createApiFactory,
+  createPlugin,
+  storageApiRef,
+  StorageApi,
+} from '@backstage/frontend-plugin-api';
+import { WorkApi, workApiRef } from '@internal/plugin-example-react';
+
+class WorkImpl implements WorkApi {
+  constructor(options: { storageApiRef: StorageApi }) {
+    /* TODO */
+  }
+  async doWork() {
+    /* TODO */
+  }
+}
+
+const workApi = createApiExtension({
+  api: workApiRef,
+  factory: () =>
+    createApiFactory({
+      api: workApiRef,
+      deps: { storageApi: storageApiRef },
+      factory: ({ storageApi }) => new WorkImpl({ storageApi }),
+    }),
+});
```

Comment:
```suggestion
const exampleWorkApi = createApiExtension({
  factory: createApiFactory({
    api: workApiRef,
    deps: { storageApi: storageApiRef },
    factory: ({ storageApi }) => new WorkImpl({ storageApi }),
  }),
});
```

The callback form is only needed if you want to access anything from the extension factory context, otherwise it's a bit verbose

Reply:
sure, i did the double form because the steps below (especially after the additions I have in the pipeline after this PR) make for a clearer diff that way. But this is fine, we can work around that